--- old/ftpd.c	2015-02-09 20:48:01.000000000 +0100
+++ new/ftpd.c	2015-04-29 18:30:00.000000000 +0200
@@ -1417,6 +1417,11 @@
             remove_callbacks (data->DATA_SOCK);
         };
     };
+    if (data->DATA_SOCK_STATUS == 'C') {
+        /* if not yet active, just wait
+           and the callback will be put in once the connection is made */
+        remove_callbacks (data->DATA_SOCK);
+    };
 };
 
 /*
