--- a/comm3705.c	2012-02-14 21:27:26.000000000 +0100
+++ b/comm3705.c	2012-02-14 21:34:07.000000000 +0100
@@ -1738,7 +1738,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA request2"); 
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
         respbuf[0] = 0x1C;
@@ -1812,7 +1812,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA request3"); 
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
         respbuf[0] = 0x1D;
@@ -1856,7 +1856,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA request4"); 
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
         respbuf[0] = 0x1C;
@@ -1923,7 +1923,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA request5"); 
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
         respbuf[0] = 0x1C;
@@ -1968,7 +1968,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA request"); 
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
 //        respbuf[0] = requestp[0];
@@ -2049,7 +2049,7 @@
                 WRMSG(HHC01020, "E", SSID_TO_LCSS(ca->dev->ssid), ca->dev->devnum, "SNA response");
                 return;
         }
-        respbuf = 4 + (BYTE*)eleptr;
+        respbuf = SIZEOF_INT_P + (BYTE*)eleptr;
 
         /* first do the ten-byte FID1 TH */
         respbuf[0] = requestp[0];
@@ -2210,7 +2210,7 @@
                 eleptr = get_bufpool(&dev->commadpt->sendq);
                 *residual=count;
                 if (eleptr) {
-                    piudata = 4 + (BYTE*)eleptr;
+                    piudata = SIZEOF_INT_P + (BYTE*)eleptr;
                     piusize = (piudata[8] << 8) + piudata[9];
                     piusize += 10;    // for FID1 TH
                     iobuf[0] = BUFPD;
